keps/sig-scheduling/2044-topology-awareness-in-kube-scheduler/README.md

```diff
@@ -0,0 +1,416 @@
+    Value int	  `json:"value"`
+}
+
+type AttributeInfo struct {
+    Name  string `json:"name"`
+    Value string `json:"value"`
+}
+
+type CostList []CostInfo
+type AttributeList []AttributeInfo
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NodeResourceTopologyList is a list of NodeResourceTopology resources
+type NodeResourceTopologyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []NodeResourceTopology `json:"items"`
+}
+
+```
+
+
+Where TopologyPolicy may have following values: none, best-effort, restricted, single-numa-node.
+The current policies of TopologyManager can't coexist together at the same time, but in future such kind of policies could appear.
+For example we can have policy for HyperThreading and it can live with NUMA policies.
+
+To use these policy names both in kube-scheduler and in kubelet, string constants of these labels should be moved from pkg/kubelet/cm/topologymanager/ and pkg/kubelet/apis/config/types.go to pkg/apis/core/types.go a one single place.
+
+## Plugin implementation details
+
+### Description of the Algorithm
+
+The algorithm which implements single-numa-node policy is following:
+
+```go
+	if qos == v1.PodQOSBestEffort {
+		return nil
+	}
+
+	zeroQuantity := resource.MustParse("0")
+	for _, container := range containers {
+		bitmask := bm.NewEmptyBitMask()
+		bitmask.Fill()
+		for resource, quantity := range container.Resources.Requests {
+			resourceBitmask := bm.NewEmptyBitMask()
+			for _, numaNode := range zones {
+				numaQuantity, ok := numaNode.Resources[resource]
+				// if can't find requested resource on the node - skip (don't set it as available NUMA node)
+				// if unfound resource has 0 quantity probably this numa node can be considered
+				if !ok && quantity.Cmp(zeroQuantity) != 0{
+					continue
+				}
+				// Check for the following:
+				// 1. set numa node as possible node if resource is memory or Hugepages (until memory manager will not be merged and
+				// memory will not be provided in CRD
+				// 2. set numa node as possible node if resource is cpu and it's not guaranteed QoS, since cpu will flow
+				// 3. set numa node as possible node if zero quantity for non existing resource was requested (TODO check topology manaager behaviour)
+				// 4. otherwise check amount of resources
+				if resource == v1.ResourceMemory ||
+					strings.HasPrefix(string(resource), string(v1.ResourceHugePagesPrefix)) ||
+					resource == v1.ResourceCPU && qos != v1.PodQOSGuaranteed ||
+					quantity.Cmp(zeroQuantity) == 0 ||
+					numaQuantity.Cmp(quantity) >= 0 {
+					resourceBitmask.Add(numaNode.NUMAID)
+				}
+			}
+			bitmask.And(resourceBitmask)
+		}
+		if bitmask.IsEmpty() {
+			// definitely we can't align container, so we can't align a pod
+			return framework.NewStatus(framework.Unschedulable, fmt.Sprintf("Can't align container: %s", container.Name))
+		}
+	}
+	return nil
+}
+```
+
+
+
+# Alternative Solution
```

Comment:
in this section, I would describe the alternative proposal, pros/cons compared to primary one and why the primary one is better. Also, this is not proposing an end-to-end alternative, only the kubelet part, if so please make that clear as well.

The proto details mentioned in the following section are not really helpful in evaluating that.

Reply:
can you discuss the following potential end-to-end alternative: 1:1 worker pod to node assignment, so apart from kubelet and daemonsets, the pod will take the whole node and the application is responsible for forking processes and assign them to NUMA cells. How feasible is this, do we have frameworks/libraries that allows applications to do that (e.g., do MPI libraries enable this)?

Reply:
> in this section, I would describe the alternative proposal, pros/cons compared to primary one and why the primary one is better. Also, this is not proposing an end-to-end alternative, only the kubelet part, if so please make that clear as well.

I will fix this and make this more coherent.

 
> can you discuss the following potential end-to-end alternative: 1:1 worker pod to node assignment, so apart from kubelet and daemonsets, the pod will take the whole node and the application is responsible for forking processes and assign them to NUMA cells. How feasible is this, do we have frameworks/libraries that allows applications to do that (e.g., do MPI libraries enable this)?

This is a very interesting point! Topology Manager can deal with the alignment of resources at container and pod scope. So if we have a pod with multiple containers, the individual processes runtime inside the containers can be assigned to a NUMA cell but I don't think we would be able to more granular than that (assign processes inside container to separate NUMA nodes) without making major changes to Kubelet.